RAT/controls.py

```diff
@@ -0,0 +1,212 @@
+                      data: dict[str, Any],
+                      expected_properties: list[str],
+                      msg: str) -> None:
+        if isinstance(data, dict) and data != {}:
+            if not all(x in expected_properties for x in data.keys()):
+                raise ValueError(msg)
+
+
+class Calculate(BaseProcedure, validate_assignment = True, extra = 'forbid'):
+    """
+    Defines the class for the calculate procedure.
+    """
+    procedure: Procedures = Field(Procedures.Calculate, frozen = True)
+     
+    @model_validator(mode = 'before')
+    @classmethod
+    def verify_inputs(cls, data: dict[str, Any]) -> dict[str, Any]:
+        expected_properties = ['parallel',
+                               'calcSldDuringFit',
+                               'resamPars',
+                               'display',
+                               'procedure']
+        msg = ("Properties that can be set for calculate are"
+               " calcSLdDuringFit, display, parallel, resamPars")
+        super().verify_inputs(data, expected_properties, msg)
+        return data
+
+
+class Simplex(BaseProcedure, validate_assignment = True, extra = 'forbid'):
+    """
+    Defines the class for the simplex procedure.
+    """
+    procedure: Procedures = Field(Procedures.Simplex, frozen = True)
+    tolX: float = Field(1e-6, gt = 0)
+    tolFun: float = Field(1e-6, gt = 0)
+    maxFunEvals: int = Field(10000, gt = 0)
+    maxIter: int = Field(1000, gt = 0)
+    updateFreq: int = -1
+    updatePlotFreq: int = -1
+
+    @model_validator(mode = 'before')
+    @classmethod
+    def verify_inputs(self, data: dict[str, Any]) -> dict[str, Any]:
+        expected_properties = ['parallel',
+                               'calcSldDuringFit',
+                               'resamPars',
+                               'display',
+                               'procedure',
+                               'tolX',
+                               'tolFun',
+                               'maxFunEvals',
+                               'maxIter',
+                               'updateFreq',
+                               'updatePlotFreq']
+        msg = ("Properties that can be set for simplex are"
+               " calcSLdDuringFit, display, maxFunEvals, maxIter,"
+               " parallel, resamPars, tolFun, tolX, updateFreq, "
+               "updatePlotFreq")
+        super().verify_inputs(data, expected_properties, msg)
+        return data
+
+
+class DE(BaseProcedure, validate_assignment = True, extra = 'forbid'):
+    """
+    Defines the class for the Differential Evolution procedure.
+    """
+    procedure: Procedures = Field(Procedures.DE, frozen = True)
+    populationSize: int = Field(20, ge = 1)
+    fWeight: float = 0.5
+    crossoverProbability: float = Field(0.8, gt = 0, lt = 1)
+    strategy: StrategyOptions = StrategyOptions.RandomWithPerVectorDither
+    targetValue: Union[int, float] = Field(1.0, ge = 1)
```

Comment:
There were some fields that had 0 as the minimum value and there were some with integers values allowed. I know that people don't explicitly specify 0 as 0.0 or 1 as 1.0. So this was done to allow both inputs to be accepted. I was doing this in the previous classes. I was type casting 4 variables.

Reply:
This strikes me as sloppy coding rather than anything else. We should establish on Monday what the types of these fields actually are and use just that type. If someone types `targetValue=5` pydantic will automatically apply `targetValue=float(targetValue)`, so that input isn't a problem at all.

Reply:
I have converted all the ones with `Union[int, float]` to float.